controllers/gcpmachine_controller_test.go

```diff
@@ -48,4 +64,171 @@ var _ = Describe("GCPMachineReconciler", func() {
+					Namespace: "default",
+				},
+				Spec: clusterv1.MachineSpec{
+					InfrastructureRef: corev1.ObjectReference{
+						APIVersion: "infrastructure.cluster.x-k8s.io/v1alpha3",
+						Kind:       "GCPMachine",
+						Name:       gcpMachine.Name,
+					},
+					Bootstrap: clusterv1.Bootstrap{
+						Data: &bootstrapData,
+					},
+					Version: &version,
+				},
+			}
+
+			Expect(k8sClient.Create(ctx, gcpMachine)).Should(Succeed())
+
+			cluster := &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "cluster",
+					Namespace: "default",
+				},
+				Status: clusterv1.ClusterStatus{
+					InfrastructureReady: true,
+				},
+			}
+
+			gcpCluster := &infrav1.GCPCluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "gcpcluster",
+					Namespace: "default",
+				},
+				Spec: infrav1.GCPClusterSpec{
+					Project: "gcp-project",
+				},
+			}
+
+			// Create the cluster scope
+			clusterScope, err = scope.NewClusterScope(scope.ClusterScopeParams{
+				Client:     fake.NewFakeClient(),
+				Logger:     log.Log,
+				Cluster:    cluster,
+				GCPCluster: gcpCluster,
+			})
+			Expect(err).To(BeNil())
+
+			// Create the machine scope
+			machineScope, err = scope.NewMachineScope(scope.MachineScopeParams{
+				Logger:     log.Log,
+				Client:     fake.NewFakeClient(),
+				Cluster:    cluster,
+				Machine:    machine,
+				GCPCluster: gcpCluster,
+				GCPMachine: gcpMachine,
+			})
+			Expect(err).To(BeNil())
+
+		})
+
+		AfterEach(func() {
+			Expect(k8sClient.Delete(ctx, machineScope.GCPMachine)).Should(Succeed())
+		})
+
+		It("should return early if gcp machine is in error state", func() {
+			machineScope.SetErrorReason(capierrors.UpdateMachineError)
+			machineScope.SetErrorMessage(errors.New("GCE error"))
+			Expect(k8sClient.Update(ctx, machineScope.GCPMachine)).Should(Succeed())
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Error state detected, skipping reconciliation"))
+		})
+
+		It("should return if infrastructure is not ready", func() {
+			clusterStatus := clusterv1.ClusterStatus{InfrastructureReady: false}
+			machineScope.Cluster.Status = clusterStatus
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Cluster infrastructure is not ready yet"))
+
+		})
+
+		It("should return if bootstrap data is not available", func() {
+			spec := clusterv1.MachineSpec{
+				Bootstrap: clusterv1.Bootstrap{
+					Data: nil,
+				}}
+			machineScope.Machine.Spec = spec
+
+			buf := new(bytes.Buffer)
+			klog.SetOutput(buf)
+
+			_, err := reconciler.reconcile(ctx, machineScope, clusterScope)
+			Expect(err).To(BeNil())
+			Expect(buf.String()).To(ContainSubstring("Bootstrap data is not yet available"))
```

Comment:
Checking for a specific log message seems a bit brittle. Do we need a better way to bubble this information up?

Reply:
there is a series of checks that happen before this call is made, so I don't think just checking for InstanceIfExists not being called would be enough replacement that the right check failed